src/python/pants/build_graph/target.py

```diff
@@ -473,7 +473,10 @@ def transitive_invalidation_hash(self, fingerprint_strategy=None, depth=0):
       def dep_hash_iter():
         for dep in self.dependencies:
           try:
-            dep_hash = dep.transitive_invalidation_hash(fingerprint_strategy, depth=depth+1)
+            if getattr(self, 'strict_deps', None):
+              dep_hash = dep.invalidation_hash(fingerprint_strategy)
```

Comment:
Unfortunately, I think that this will break assumptions in a lot of Tasks. Whether invalidation is transitive or direct is currently a task-level concern. And in particular, "strict_deps" needs to only apply to tasks that explicitly ask for it.

As an example: `zinc` will use strict deps, and only pass strict deps to the compiler. But `junit` always uses the full transitive graph.

So I think that (with the current task API) it is necessary for this to be a change to this argument: https://github.com/pantsbuild/pants/blob/1e45167fab9a30f32207d025d91d2a071c8981fe/src/python/pants/task/task.py#L270 ... I definitely don't love that tasks are responsible for specifying this, but it's the way of things currently.